cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -99,82 +111,145 @@ protected void startUp() throws Exception {
       .build();
 
     httpService.startAndWait();
-    running = true;
+
+    executor = Executors.newSingleThreadExecutor();
   }
 
   @Override
   protected void shutDown() throws Exception {
     httpService.stopAndWait();
+    shutdownExecutor();
+  }
+
+  private void executeWorkflowAction(ApplicationSpecification appSpec, WorkflowActionNode node,
```

Comment:
You are already inside `WorkflowDriver`, you don't need to repeat `workflow` in private method. Just call it `executeAction`